src/qibolab/_core/parameters.py

```diff
@@ -202,3 +218,118 @@ def replace(self, update: Update) -> "Parameters":
             _setvalue(d, path, val)
 
         return self.model_validate(d)
+
+
+QubitMap = dict[QubitId, Qubit]
+InstrumentMap = dict[InstrumentId, Instrument]
+
+
+class Hardware(TypedDict):
+    """Part of the platform that specifies the hardware configuration."""
+
+    instruments: InstrumentMap
+    qubits: QubitMap
+    couplers: NotRequired[QubitMap]
+
+
+def _gate_channel(qubit: Qubit, gate: str) -> str:
+    """Default channel that a native gate plays on."""
+    if gate in ("RX", "RX90", "CNOT"):
+        return qubit.drive
+    if gate == "RX12":
+        return qubit.drive_qudits[(1, 2)]
+    if gate == "MZ":
+        return qubit.acquisition
+    if gate in ("CP", "CZ", "iSWAP"):
+        return qubit.flux
+
+
+def _gate_sequence(qubit: Qubit, gate: str) -> Native:
+    """Default sequence corresponding to a native gate."""
+    channel = _gate_channel(qubit, gate)
+    pulse = Pulse(duration=0, amplitude=0, envelope=Rectangular())
+    if gate != "MZ":
+        return Native([(channel, pulse)])
+
+    return Native(
+        [(channel, Readout(acquisition=Acquisition(duration=0), probe=pulse))]
+    )
+
+
+def _pair_to_qubit(pair: str, qubits: QubitMap) -> Qubit:
+    """Get first qubit of a pair given in ``{q0}-{q1}`` format."""
+    q = tuple(pair.split("-"))[0]
+    try:
+        return qubits[q]
+    except KeyError:
+        return qubits[int(q)]
+
+
+def _native_builder(cls, qubit: Qubit, natives: set[str]) -> NativeContainer:
+    """Build default native gates for a given qubit or pair.
+
+    In case of pair, ``qubit`` is assumed to be the first qubit of the pair,
+    and a default pulse is added on that qubit, because at this stage we don't
+    know which qubit is the high frequency one.
+    """
+    return cls(
+        **{
+            gate: _gate_sequence(qubit, gate)
+            for gate in cls.model_fields.keys() & natives
+        }
+    )
+
+
+def _channel_config(id: ChannelId, channel: Channel) -> dict[ChannelId, Config]:
+    """Default configs correspondign to a channel."""
+    if isinstance(channel, DcChannel):
+        return {id: DcConfig(offset=0)}
+    if isinstance(channel, AcquisitionChannel):
+        return {id: AcquisitionConfig(delay=0, smearing=0)}
+    if isinstance(channel, IqChannel):
+        configs = {id: IqConfig(frequency=0)}
+        if channel.lo is not None:
+            configs[channel.lo] = OscillatorConfig(frequency=0, power=0)
+        if channel.mixer is not None:
+            configs[channel.mixer] = IqMixerConfig(frequency=0, power=0)
+        return configs
+    return {id: Config()}
+
+
+class ParametersBuilder(Model):
+    """Generates default ``Parameters`` for a given platform hardware
+    configuration."""
+
+    hardware: Hardware
+    natives: set[str] = Field(default_factory=set)
+    pairs: list[str] = Field(default_factory=list)
+
+    def build(self) -> Parameters:
```

Comment:
Changed to function in 0375df894c744868e0178fdc124aadab26aeb928, just because it is slightly simpler, even though I am also not against the builder, as it gives us runtime type checking and some casting for free.

> or even a constructor of `Parameters` (but maybe I'd avoid this last option, to keep it more modular).

Out of curiosity, if we are going with function, what is the problem with the constructor? From the user side, it seems easier to remember:
```py
from qibolab import Parameters

parameters = Parameters.build(...)
```
than
```py
from qibolab import initialize_parameters

parameters = initialize_parameters(...)
```
It is just that in the first case we have to make `Parameters` public, as currently it is not, however we have to make the initializer public anyway even in the second case.

Reply:
To be fair, we could even make it a constructor, for the time being. But I consider classes in Python not very scalable, mostly because they are bound to have all definitions in a single block (which has also to live in a single module), and after that we should start defining functions, at least internally, to then alias them with class' methods.

In general, the `classmethod` and `staticmethod` as well are not a bad idea, since they allow placing some functions in a certain scope. But you're not completely free to play with scopes, since you're *constrained* by the single block.

In general, I'm trying to split data and their manipulation, to isolate more the operations. Thus, I'm pushing for dataclasses (and similar) to hold the information, but often with little to no methods - using functions to implement the operations.

With "isolating" I mean limiting the data access, and limit even more mutation, whenever possible. Which is a bit against how classes are often used.
While I consider this an improvement, I'm not at all against every aspect of Python classes. Generating scopes is a good one - just limited.